### ISSUE TITLE
Eval mode forward disturbs bf16 fsdp checkpointing

### ISSUE THREAD

**Issue by daviswer** [pytorch/pytorch#104770]
### 🐛 Describe the bug

When training an FSDP/HSDP model in bf16, performing a forward pass in eval mode, without a corresponding backward pass, at any point, will cause future checkpoints to load incorrectly. The behavior of the trained model is unaffected, but when loading from checkpoint, the new model will behave very differently from the old one on the same input data. 

The inconsistent behavior can be eliminated by:

- Skipping eval mode when performing the forward pass without a backward pass
- Adding a backward pass to the step in question
- Working in 32-bit precision
- Switching model to DDP
- Putting the new model, loaded from checkpoint, in eval mode

While admittedly a very specific edge case, this is of interest in cases where we'd want to produce a model "signature" associated with each checkpoint, for correctness checking purposes, by doing a forward pass on a small canonical input. Ironically, the correctness check is destroying the correctness of the checkpoints (and more broadly, a forward pass before the training loop even starts should not affect long-term checkpointing behavior). 

Code to reproduce this is below - using torchrun I create a two-layer embedding / prediction head network, do a dummy forward pass in eval mode, then train it to map each input index to the one higher, and save to checkpoint. A second model, loaded from that checkpoint, produces very different logits from the first, on the same inputs. 

```
import functools
import math
import os

import torch
import torch.distributed as dist
import torch.nn as nn
from torch.distributed._shard.checkpoint import (
    FileSystemReader,
    FileSystemWriter,
    load_state_dict,
    save_state_dict,
)
from torch.distributed.checkpoint.default_planner import (
    DefaultLoadPlanner,
    DefaultSavePlanner,
)
from torch.distributed.fsdp import BackwardPrefetch, CPUOffload, FullStateDictConfig
from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
from torch.distributed.fsdp import MixedPrecision, ShardingStrategy, StateDictType
from torch.distributed.fsdp.wrap import enable_wrap, transformer_auto_wrap_policy, wrap

savepath = "/data/output/davis-1b/TEMP/"

local_rank = int(os.getenv("LOCAL_RANK"))

torch.manual_seed(42 + local_rank)
torch.cuda.manual_seed_all(42 + local_rank)

mp_policy = MixedPrecision(
    param_dtype=torch.bfloat16,
    reduce_dtype=torch.bfloat16,
    buffer_dtype=torch.bfloat16,
)

wrapping_policy = functools.partial(transformer_auto_wrap_policy, transformer_layer_cls={})

model_sharding_strategy = ShardingStrategy.FULL_SHARD

torch.cuda.set_device(int(os.getenv("RANK")))
dist.init_process_group("nccl")

v = 2048
d = 1024

model = nn.Sequential(nn.Embedding(v, d), nn.Linear(d, v))

model = FSDP(
    model,
    auto_wrap_policy=wrapping_policy,
    mixed_precision=mp_policy,
    sharding_strategy=model_sharding_strategy,
    device_id=local_rank,
    limit_all_gathers=True,
    use_orig_params=True,
)

optimizer = torch.optim.AdamW(model.parameters(), lr=0.0001)
criterion = nn.CrossEntropyLoss()

# THIS TRIGGERS THE FUTURE CHECKPOINT LOAD FAILURE
inp = torch.arange(256, device=local_rank).unsqueeze(0)
model.eval()
out = model(inp)
model.train()

# TRAINING LOOP
for i in range(100):
    inp = torch.randint(v, (8, 256), device=local_rank)
    label = inp.add(1) % v
    optimizer.zero_grad()
    pred = model(inp)
    loss = criterion(pred.view(-1, pred.size(-1)), label.view(-1))
    loss.backward()
    optimizer.step()
    if local_rank == 0 and (i + 1) % 10 == 0:
        print(f"Step {i+1}, loss {loss.item()}")

# SAVE CHECKPOINT
with FSDP.state_dict_type(model, StateDictType.SHARDED_STATE_DICT):
    model_state = model.state_dict()
    writer = FileSystemWriter(savepath, single_file_per_rank=True)
    save_state_dict({"model_state": model_state}, storage_writer=writer, planner=DefaultSavePlanner())

# BUILD SECOND MODEL
model2 = nn.Sequential(nn.Embedding(v, d), nn.Linear(d, v))
model2 = FSDP(
    model2,
    auto_wrap_policy=wrapping_policy,
    mixed_precision=mp_policy,
    sharding_strategy=model_sharding_strategy,
    device_id=local_rank,
    limit_all_gathers=True,
    use_orig_params=True,
)

# LOAD CHECKPOINT INTO MODEL 2
with FSDP.state_dict_type(model2, StateDictType.SHARDED_STATE_DICT):
    state_dict = model2.state_dict()
    state_ckp = {"model_state": state_dict}
    load_state_dict(
        state_dict=state_ckp,
        storage_reader=FileSystemReader(savepath),
        planner=DefaultLoadPlanner(),
    )
    model2.load_state_dict(state_ckp["model_state"])
    model2.to(local_rank)

# COMPARE OUTPUTS
n = 10
probe = torch.arange(n, device=local_rank).unsqueeze(0)
p = probe.reshape(-1).tolist()
model.train()
model2.train()
out1 = model(probe)[0, :n, :6].reshape(-1).tolist()
out2 = model2(probe)[0, :n, :6].reshape(-1).tolist()
if local_rank == 0:
    for i in range(n):
        print("Key:", p[i])
        print("Model 1:", out1[i * 6 : i * 6 + 6])
        print("Model 2:", out2[i * 6 : i * 6 + 6])
        print()
```

Expected behavior is that the model saving the checkpoint and the model loaded from checkpoint should return the same outputs for the same inputs. Instead we get (printing the first 6 logits for tokens 0-9):
```
Key: 0
Model 1: [0.318359375, 7.125, 0.0546875, 0.375, 0.142578125, -1.390625]
Model 2: [0.23828125, -0.01806640625, 0.068359375, 0.1953125, 0.318359375, -0.9921875]

Key: 1
Model 1: [0.78125, -0.3984375, 5.96875, -0.42578125, -0.63671875, 0.349609375]
Model 2: [0.48046875, -0.41796875, -0.91796875, -0.400390625, -0.2353515625, 0.158203125]

Key: 2
Model 1: [-0.32421875, 0.6796875, 0.59375, 8.125, 0.216796875, -0.53125]
Model 2: [-0.5, -0.0233154296875, 0.59765625, 0.98046875, -0.150390625, -0.28125]

Key: 3
Model 1: [-0.75390625, -0.7265625, -1.5859375, -0.1943359375, 7.40625, -1.2734375]
Model 2: [-0.5859375, -0.486328125, -1.2890625, -0.69140625, 0.259765625, -1.2890625]

Key: 4
Model 1: [-0.3125, -0.6875, 0.5390625, 0.042724609375, -0.248046875, 7.28125]
Model 2: [-0.5390625, -0.24609375, 0.259765625, 0.24609375, -0.1435546875, -0.3515625]

Key: 5
Model 1: [-0.1884765625, 0.216796875, -0.142578125, -0.474609375, 0.392578125, 0.318359375]
Model 2: [0.416015625, -0.054931640625, -0.162109375, -0.427734375, 0.392578125, 0.1181640625]

Key: 6
Model 1: [0.5, -0.453125, 1.21875, 0.30859375, 0.265625, 0.408203125]
Model 2: [0.63671875, -0.333984375, 0.73046875, -0.162109375, -0.255859375, 0.20703125]

Key: 7
Model 1: [0.328125, 0.6328125, 0.275390625, 0.66015625, 0.07421875, 1.03125]
Model 2: [0.173828125, 0.1826171875, 0.2041015625, 0.8359375, 0.234375, 0.921875]

Key: 8
Model 1: [-0.3125, -0.90625, -0.43359375, 0.94921875, -0.80078125, -1.171875]
Model 2: [-0.33984375, -0.4296875, -0.419921875, 0.146484375, -0.6640625, -1.3046875]

Key: 9
Model 1: [-0.7109375, 0.078125, -0.78515625, -0.6484375, 0.734375, -0.337890625]
Model 2: [-0.703125, -0.203125, -0.52734375, -0.63671875, 0.6875, 0.10302734375]
```

### Versions

```
PyTorch version: 2.1.0.dev20230612+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Debian GNU/Linux 11 (bullseye) (x86_64)
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.31

Python version: 3.10.10 (main, Mar 21 2023, 18:45:11) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-4.18.0-372.19.1.el8_6.x86_64-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA A100-SXM4-80GB
GPU 1: NVIDIA A100-SXM4-80GB
GPU 2: NVIDIA A100-SXM4-80GB
GPU 3: NVIDIA A100-SXM4-80GB

Nvidia driver version: 515.48.07
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   46 bits physical, 48 bits virtual
CPU(s):                          80
On-line CPU(s) list:             0-79
Thread(s) per core:              2
Core(s) per socket:              20
Socket(s):                       2
NUMA node(s):                    2
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           85
Model name:                      Intel Xeon Processor (Cascadelake)
Stepping:                        5
CPU MHz:                         2399.998
BogoMIPS:                        4799.99
Virtualization:                  VT-x
Hypervisor vendor:               KVM
Virtualization type:             full
L1d cache:                       1.3 MiB
L1i cache:                       1.3 MiB
L2 cache:                        160 MiB
L3 cache:                        32 MiB
NUMA node0 CPU(s):               0-39
NUMA node1 CPU(s):               40-79
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology cpuid tsc_known_freq pni pclmulqdq vmx ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch cpuid_fault invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves arat pku ospke avx512_vnni md_clear arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.25.0rc1
[pip3] pytorch-triton==2.1.0+9820899b38
[pip3] torch==2.1.0.dev20230612+cu118
[pip3] torchvision==0.16.0.dev20230612+cu118
[conda] numpy                     1.25.0rc1                pypi_0    pypi
[conda] pytorch-triton            2.1.0+9820899b38          pypi_0    pypi
[conda] torch                     2.1.0.dev20230612+cu118          pypi_0    pypi
[conda] torchvision               0.16.0.dev20230612+cu118          pypi_0    pypi
```

cc @ezyang @gchanan @zou3519 @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu

**Comment by rohan-varma** [comment]
adding hi-pri due to silent correctness issue
